### PR TITLE
[trainer] feat: support HSDP for expert parallel

### PR DESCRIPTION
**Pull request by wuxibin89**
### What does this PR do?

Support HSDP for extra parallel. For example, world_size=1024
- dense param device_mesh: `(dp_replicate, dp_shard_sp)=(4, 256)`
- ep_size=8, expert parallel device_mesh: `(ep_replicate, ep_fsdp, ep)=(4, 32, 8)`

### Test

1. Hardware: 2*8 A100
2. Test script
```bash
#!/usr/bin/env bash
set -xeuo pipefail

# set dist args
nproc_per_node=${ARNOLD_WORKER_GPU}
export NNODES=${ARNOLD_WORKER_NUM}
export NODE_RANK=${ARNOLD_ID}
master_addr="ARNOLD_WORKER_0_HOST"
export MASTER_ADDR=${!master_addr}
master_port="ARNOLD_WORKER_0_PORT"
export master_port=${!master_port}
ports=(`echo $master_port | tr ',' ' '`)
export MASTER_PORT=${ports[0]}
echo "[nproc_per_node: ${nproc_per_node}]"
echo "[nnodes: ${NNODES}]"
echo "[node_rank: ${NODE_RANK}]"
echo "[master_addr: ${MASTER_ADDR}]"
echo "[master_port: ${MASTER_PORT}]"

# Setting 1:
# (ep_replicate, ep_fsdp, ep)=(2, 2, 4)
# dp_replicate_size=2
# dp_shard_size=8
# ep_size=4

# Setting 2:
# (ep_replicate, ep_fsdp, ep)=(2, 1, 8)
# dp_replicate_size=2
# dp_shard_size=8
# ep_size=8

# Setting 3:
# (ep_fsdp, ep)=(4, 4)
dp_replicate_size=1
dp_shard_size=16
ep_size=4

export VEOMNI_VERBOSITY=warn

bash train.sh tasks/train_text.py configs/text/qwen3.yaml \
    --model.model_path $DATA_ROOT/model/Qwen3-30B-A3B-Instruct-2507 \
    --model.ops_implementation.moe_implementation fused_triton \
    --data.train_path $DATA_ROOT/dataset/allenai/tulu-3-sft-mixture/data/train-00000-of-00006.parquet \
    --train.wandb.project wuxibin_veomni \
    --train.accelerator.fsdp_config.fsdp_mode fsdp2 \
    --train.accelerator.dp_replicate_size $dp_replicate_size \
    --train.accelerator.dp_shard_size $dp_shard_size \
    --train.accelerator.ep_size $ep_size \
    --train.init_device meta \
    --train.global_batch_size 16 \
    --train.max_steps 5 \
    --train.num_train_epochs 1
```
3. Test result
- Setting 1: (ep_replicate, ep_fsdp, ep)=(2, 2, 4)
```text
Epoch 1/1:   0%|          | 0/5 [00:12<?, ?it/s, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.79, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:12<00:50, 12.65s/it, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.79, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:15<00:50, 12.65s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.69, lr: 0.00] 
Epoch 1/1:  40%|████      | 2/5 [00:15<00:20,  6.78s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.69, lr: 0.00]
Epoch 1/1:  40%|████      | 2/5 [00:17<00:20,  6.78s/it, total_loss: 1.42, foundation_loss: 1.42, grad_norm: 2.69, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:17<00:09,  4.88s/it, total_loss: 1.42, foundation_loss: 1.42, grad_norm: 2.69, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:20<00:09,  4.88s/it, total_loss: 1.38, foundation_loss: 1.38, grad_norm: 2.90, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:20<00:03,  3.98s/it, total_loss: 1.38, foundation_loss: 1.38, grad_norm: 2.90, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:23<00:03,  3.98s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.50, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:23<00:00,  3.49s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.50, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:23<00:00,  4.63s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.50, lr: 0.00]
```
- Setting 2: (ep_replicate, ep_fsdp, ep)=(2, 1, 8)
```text
Epoch 1/1:   0%|          | 0/5 [00:15<?, ?it/s, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.80, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:15<01:03, 15.89s/it, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.80, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:18<01:03, 15.89s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.80, lr: 0.00] 
Epoch 1/1:  40%|████      | 2/5 [00:18<00:24,  8.15s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.80, lr: 0.00]
Epoch 1/1:  40%|████      | 2/5 [00:21<00:24,  8.15s/it, total_loss: 1.43, foundation_loss: 1.43, grad_norm: 2.75, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:21<00:11,  5.59s/it, total_loss: 1.43, foundation_loss: 1.43, grad_norm: 2.75, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:23<00:11,  5.59s/it, total_loss: 1.37, foundation_loss: 1.37, grad_norm: 2.93, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:23<00:04,  4.38s/it, total_loss: 1.37, foundation_loss: 1.37, grad_norm: 2.93, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:26<00:04,  4.38s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.52, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:26<00:00,  3.74s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.52, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:26<00:00,  5.26s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.52, lr: 0.00]
```
- Setting 3: (ep_fsdp, ep)=(4, 4)
```text
Epoch 1/1:   0%|          | 0/5 [00:17<?, ?it/s, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.79, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:17<01:10, 17.59s/it, total_loss: 1.94, foundation_loss: 1.94, grad_norm: 16.79, lr: 0.00]
Epoch 1/1:  20%|██        | 1/5 [00:20<01:10, 17.59s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.80, lr: 0.00] 
Epoch 1/1:  40%|████      | 2/5 [00:20<00:27,  9.02s/it, total_loss: 1.52, foundation_loss: 1.52, grad_norm: 6.80, lr: 0.00]
Epoch 1/1:  40%|████      | 2/5 [00:23<00:27,  9.02s/it, total_loss: 1.42, foundation_loss: 1.42, grad_norm: 2.85, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:23<00:12,  6.15s/it, total_loss: 1.42, foundation_loss: 1.42, grad_norm: 2.85, lr: 0.00]
Epoch 1/1:  60%|██████    | 3/5 [00:26<00:12,  6.15s/it, total_loss: 1.36, foundation_loss: 1.36, grad_norm: 2.69, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:26<00:04,  4.90s/it, total_loss: 1.36, foundation_loss: 1.36, grad_norm: 2.69, lr: 0.00]
Epoch 1/1:  80%|████████  | 4/5 [00:29<00:04,  4.90s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.48, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:29<00:00,  4.14s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.48, lr: 0.00]
Epoch 1/1: 100%|██████████| 5/5 [00:29<00:00,  5.83s/it, total_loss: 1.29, foundation_loss: 1.29, grad_norm: 1.48, lr: 0.00]
```
### API and Usage Example

No API change.

### CODE REVIEW

File: veomni/checkpoint/dcp_checkpointer.py [hunk 1]
```diff
@@ -275,45 +293,97 @@ def get_state_dict_with_extra_parallel_dim_preprocess(self, state_dict, action):
...
+        if fsdp_mesh.ndim == 3:
+            placements = [Replicate(), Shard(fsdp_shard_dim), Shard(ep_shard_dim)]
+        else:
+            placements = [Shard(fsdp_shard_dim), Shard(ep_shard_dim)]
+        dtensor = DTensor.from_local(orgin_tensor._local_tensor, device_mesh=fsdp_mesh, placements=placements)
     elif torch.is_tensor(orgin_tensor):
         # If there is no FSDP but only ExtraParallel
-        dtensor = DTensor.from_local(orgin_tensor, device_mesh=extra_parallel_fsdp_mesh, placements=[Shard(0)])
+        if extra_parallel_fsdp_mesh.ndim == 2:
+            placements = [Replicate(), Shard(fsdp_shard_dim)]
```
Comment (Luosuu) [inline veomni/checkpoint/dcp_checkpointer.py]: For the pure ExtraParallel path there is no FSDP shard, so this should use ep_shard_dim rather than fsdp_shard_dim. Otherwise pure EP/emb checkpoints can be restored with the wrong sharding axis.

File: veomni/distributed/torch_parallelize.py [hunk 2]
```diff
@@ -233,14 +233,19 @@ def parallelize_model_fsdp2(
     extra_parallel_fsdp_kwargs = {}
     for para in parallel_state.extra_parallel_names:
         if parallel_state.extra_parallel_enabled(para):
-            para_fsdp_mesh = parallel_state.extra_parallel_fsdp_device_mesh[para][f"{para}_fsdp"]
+            para_mesh = parallel_state.extra_parallel_fsdp_device_mesh[para]
+            para_mesh_dim_names = para_mesh.mesh_dim_names
+            # exclude para_size dimension:
+            # - (ep_fsdp, ep) -> (ep_fsdp,)
+            # - (ep_replicate, ep_fsdp, ep) -> (ep_replicate, ep_fsdp)
+            para_fsdp_mesh = para_mesh[para_mesh_dim_names[:-1]]
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py:241]: This HSDP mesh change exposes a NPU-only grad scaling issue below: with dp_replicate=2, ep=4 the NPU test gets expert grad 0.125 vs expected 0.25, so the NPU reduce-scatter divide factor appears to include the replicate dim.

File: veomni/distributed/parallel_state.py [hunk 3]
```diff
@@ -595,14 +570,30 @@ def init_parallel_state(
...
+
+            # NOTE: Support HSDP for extra parallel. For example, world_size=1024
+            # - dense param device_mesh: (dp_replicate, dp_shard_sp)=(4, 256)
+            # - ep_size=8, expert parallel device_mesh: (ep_replicate, ep_fsdp, ep)=(4, 32, 8)
+            param_mesh_shape, para_mesh_dim_names = [], []
+            if dp_replicate_size > 1:
+                param_mesh_shape.append(dp_replicate_size)
+                para_mesh_dim_names.append(f"{para_name}_replicate")
+            dp_shard_sp_size = device_mesh["dp_shard_sp"].size()
+            assert dp_shard_sp_size % para_size == 0, (
```
Comment (Luosuu) [inline veomni/distributed/parallel_state.py:585]: This limits EP to the dp_shard_sp side of HSDP. A config like 32 cards with ep_size=32, dp_replicate=2, dp_shard=16 will fail this assert because 16 % 32 != 0, so EP cannot span HSDP replica groups yet.

Comment (Luosuu) [reply veomni/distributed/parallel_state.py:585]: If this is on purpose, we can make a note here

Comment (wuxibin89) [reply veomni/distributed/parallel_state.py:585]: Yes it's designed on purpose that `ep` group should fit into a `dp_shard` group. In practice, ep_size is almost never greater than 32(typically 8/16), while a dp_shard size is commonly 128/256.

Will add a note to clarify.

Comment (wuxibin89) [reply veomni/distributed/parallel_state.py:585]: supplement note above

File: veomni/checkpoint/dcp_checkpointer.py (hunk 1 above)
Comment (wuxibin89) [reply veomni/checkpoint/dcp_checkpointer.py]: Left pure ExtraParallel without FSDp shard unchanged.

File: veomni/distributed/torch_parallelize.py (hunk 2 above)
Comment (wuxibin89) [reply veomni/distributed/torch_parallelize.py:241]: This bug has been fixed in torch_npu>=2.9.0, skip grad_norm+hdsp unit tests under torch_npu<2.9.0.

**Review by Luosuu — APPROVED** [review]
